sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiLoads.java

```diff
@@ -354,25 +355,34 @@ public WriteResult expandUntriggered(
                 rowUpdateFn,
                 badRecordRouter));
 
+    PCollection<KV<DestinationT, StorageApiWritePayload>> successfulConvertedRows =
+        convertMessagesResult.get(successfulConvertedRowsTag);
+
+    if (numShards > 0) {
```

Comment:
I'm worried about this being subtly backwards incompatible. There may well be users who set this parameter today in batch (e.g. because they simply reuse their streaming code in batch) and will suddenly find their job much slower after upgrading Beam versions.

Reply:
@razvanculea correct me if I'm wrong but nobody is running batch with numStorageWriteApiStreams set and storage write api method.

Reply:
This gives more control on the batch behaviour, so it should be an improvement vs today. Because the solution when limits are hit is to change the graph by adding a redistribute step before the write in batch (as you rightly noticed in the previous comment) this step is not needed in streaming (thus making 2 different pipelines batch vs streaming).


(Setting it to 0 has the same effect as before)
Setting it to >0, will have similar effect in batch and streaming, and this is difference vs now, when in streaming it has effect, but none in batch.
If the parameter is used for streaming then it means that the pipeline behaves well with this parameter, and should still behave well in batch (same limits on number of connections apply).
The difference when today this is not set for batch is that for some batch jobs it will work because the parallelism is under the API limits, and on others (or simply re-running the same or similar batch jobs) will hit the limits.

I'm for a more consistent behaviour between steaming and batch, and I do understand that this might have an effect on batch jobs with this parameter set and ignored.

Reply:
> @razvanculea correct me if I'm wrong but nobody is running batch with numStorageWriteApiStreams set and storage write api method.

can't check because if this is set today, in batch it does nothing

Reply:
With this PR, those that might unknowingly set this shard param in batch, will see a change (because there will be redistribute that matches the number of shards set), and this change should impact the result of the jobs (they will still succeed) .
There will an extra step and the batch will be consistent with the job graph in streaming.